src/category_theory/limits/shapes/products.lean

```diff
@@ -142,6 +142,17 @@ from a family of morphisms between the factors.
 abbreviation pi.map {f g : β → C} [has_product f] [has_product g]
   (p : Π b, f b ⟶ g b) : ∏ f ⟶ ∏ g :=
 lim_map (discrete.nat_trans (λ X, p X.as))
+
+instance pi.map_mono {f g : β → C} [has_product f] [has_product g]
+  (p : Π b, f b ⟶ g b) [Π i, mono (p i)] : mono $ pi.map p :=
+{ right_cancellation := λ c α₁ α₂ eq1,
+  begin
+    ext ⟨j⟩, dsimp,
+    have eq2 := eq_whisker eq1 (pi.π _ j),
+    rwa [category.assoc, lim_map_π, category.assoc, lim_map_π, ←category.assoc, ←category.assoc,
+      discrete.nat_trans_app, cancel_mono] at eq2,
+  end }
```

Comment:
I think it could be worth it to formulate this for general limits (you could even generalize this to `is_limit.map`):
```lean
instance lim_map_mono {F G : J ⥤ C} [has_limit F] [has_limit G] (α : F ⟶ G)
  [∀ j, mono (α.app j)] : mono (lim_map α) :=
⟨λ Z u v h, limit.hom_ext $ λ j, (cancel_mono (α.app j)).1 $ by simpa using h =≫ limit.π _ j⟩
```
[Fun fact: if `C` has all limits of shape `J`, then the limit functor is a right adjoint, and hence preserves monomorphisms, so we get the above under the slightly weaker assumption that `mono α` instead of `∀ j, mono (α.app j`. When `C` has pullbacks, the two conditions are equivalent.]

Specializing this to products is slightly awkard but possible:
```lean
instance pi.map_mono {f g : β → C} [has_product f] [has_product g]
  (p : Π b, f b ⟶ g b) [Π i, mono (p i)] : mono $ pi.map p :=
begin
  haveI : ∀ j, mono
    ((@discrete.nat_trans _ _ _ (discrete.functor f) (discrete.functor g) (λ X, p X.as)).app j),
  { dsimp, apply_instance },
  apply_instance
end
```

Reply:
Do we have the fact that limit functor is a right adjoint in mathlib. If not, I will try to write one, it seems a useful result.

Reply:
There is a version of this in #15859, but I don't think that this can be used to show `is_right_adjoint lim`.

Reply:
I will postpone this PR by labelling it `blocked-by-other-PR` to see if I can get adjointness of `lim` working